doc/design/request_response_communication.md

```diff
@@ -0,0 +1,102 @@
+
+#### Typed API
+
+![typed API](diagrams/request_response_typed_api.svg)
+
+#### Untyped API
+
+![untyped API](diagrams/request_response_untyped_api.svg)
+
+#### Client Port
+
+![client port](diagrams/request_response_client_port.svg)
+
+The `ClientPortData` is located in the shared memory and contain only the data but no methods to access them.
+`ClientPortUser` is the class providing the methods for the user access and `ClientPortRouDi` provides the
+interface RouDi needs to connect the client to the server and to cleanup the port resources.
+
+#### Server Port
+
+![server port](diagrams/request_response_server_port.svg)
+
+Similar to the Client Port, the Server Port has `ServerPortData` which is located in the shared memory and contain only the data but no methods to access them.
+`ServerPortUser` is the class providing the methods for the user access and `ServerPortRouDi` provides the
+interface RouDi needs to connect the client to the server once the server offers its service and to cleanup the port resources.
+
+#### Request/Response Header
+
+![rpc header](diagrams/request_response_header.svg)
+
+Since request and response need to encode different meta-information, we also need different header for the messages.
+The common data is aggregated in `RpcBaseHeader` which contains a `RelativePointer` to the `ClientChunkQueueData_t` and a sequence ID.
+The `ClientChunkQueueData_t` pointer is used to identify the queue which receives the response.
+It must not be used to send the response since the lifetime might already have ended at the time
+the response is sent but used as identifier for `ChunkDistributor::deliverToQueue`.
+The sequence ID is used to match a response to a specific request if multiple requests are invoke asynchronously.
+Depending on the client and server options, a request might be dropped or a server could have a worker pool
+which results in sending the responses in a different order than the request were received.
+The sequence ID must be set by the user and also checked by the user on response.
+The `RequestHeader` has also the option to specify a message as fire and forget, which means it won't get a response to this request.
+
+#### Client/Server Options
+
+![client and server options](diagrams/request_response_options.svg)
+
+The client and server options can be used to control certain aspects of the clients and servers.
+Beside setting the capacity of the queues and defining whether a client should be connected and a server offering on creation,
+the behaviour for a slow client and server can be defined.
+A client can ask a server to block if its response queue is full.
+The server will obey if the corresponding `ClientTooSlowPolicy` is set to `WAIT_FOR_CLIENT`.
+If the options don't match, the client will not be connected to the server.
+
```

Comment:
And the server can ask the client to block if the request queue is full? Therefore the `ServerTooSlowPolicy`?

Reply:
The Server doesn't ask the client, since its all decided on the client side but the option is called `ServerTooSlowPolicy`